File: Blind75/200. Number of Islands/200. Number of Islands.cpp
class Solution {
public:
    int count = 0;
    int R,C;
    vector<pair<int,int>> dir={{0,1},{0,-1},{1,0},{-1,0}};
    int numIslands(vector<vector<char>>& grid) {
        int R = grid.size();
        int C = grid[0].size();

        for(int i=0; i<R; i++){
            for(int j=0; j<C; j++){
                
                if(grid[i][j]=='1'){
                    dfs(grid,i,j);
                    count++;
                }
            }
        }

        return count;
    }

    void dfs(vector<vector<char>>& grid, int r, int c){
        //base case
        if(r<0 || r>=R || c<0 || c>=C || grid[r][c]!='1') return;
        grid[r][c] = '0';
        /* for(int i=0; i<4; i++){
            int nr = r+dir[i].first;
            int nc = r+dir[i].second;
            dfs(grid,nr,nc);
        } */
        dfs(grid,r+1,c);
        dfs(grid,r-1,c);
        dfs(grid,r,c+1);
        dfs(grid,r,c-1);
    }


};

File: Blind75/200. Number of Islands/200. Number of Islands.cs
public class Solution {
    public int NumIslands(char[][] grid) {
        int count = 0;
        int R = grid.Length;
        int C = grid[0].Length;

        for(int i=0; i<R; i++){
            for(int j=0; j<C; j++){
                if(grid[i][j]=='1'){
                    count++;
                    dfs(ref grid, i, j);
                }
            }
        }
        return count;
    }
    public void dfs(ref char[][]grid, int r, int c){
        //base case
        if(r<0 || r>=grid.Length || c<0 || c>=grid[0].Length || grid[r][c]!='1') return;
        grid[r][c] = '0';
        dfs(ref grid, r+1, c);
        dfs(ref grid, r-1, c);
        dfs(ref grid, r, c+1);
        dfs(ref grid, r, c-1);
    }
}
